ocaml/testsuite/tests/typing-modes/modes.ml

```diff
@@ -232,3 +232,56 @@ Warning 250 [redundant-modality]: This global modality is redundant.
 
 type r = { global_ x : string; }
 |}]
+
+(* patterns *)
+
+let foo ?(local_ x @ unique once = 42) () = ()
+[%%expect{|
+val foo : ?x:local_ unique_ once_ int -> unit -> unit = <fun>
+|}]
+
+let foo ?(local_ x : _ @@ unique once = 42) () = ()
+[%%expect{|
+val foo : ?x:local_ unique_ once_ int -> unit -> unit = <fun>
+|}]
+
+let foo ?(local_ x : 'a. 'a -> 'a @@ unique once) = ()
+[%%expect{|
+Line 1, characters 17-48:
+1 | let foo ?(local_ x : 'a. 'a -> 'a @@ unique once) = ()
+                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: Optional parameters cannot be polymorphic
+|}]
+
+let foo ?x:(local_ (x,y) @ unique once = (42, 42)) () = ()
+[%%expect{|
+val foo : ?x:local_ unique_ once_ int * int -> unit -> unit = <fun>
+|}]
+
+let foo ?x:(local_ (x,y) : _ @@ unique once = (42, 42)) () = ()
+[%%expect{|
+val foo : ?x:local_ unique_ once_ int * int -> unit -> unit = <fun>
+|}]
+
+let foo ?x:(local_ (x,y) : 'a.'a->'a @@ unique once) () = ()
+[%%expect{|
+Line 1, characters 19-36:
+1 | let foo ?x:(local_ (x,y) : 'a.'a->'a @@ unique once) () = ()
+                       ^^^^^^^^^^^^^^^^^
+Error: Optional parameters cannot be polymorphic
+|}]
+
+(* note: Legacy mode syntax is not parsed for patterns *)
+
+(* CR zqian: currently all patterns, except those directly as function
+   arguments, DO NOT pick up modes during type checking. This should be fixed in
+   another PR. Here, we test that they at least parse. *)
```

Comment:
I'm not inclined to support this syntax until it means something. Can we reject this in type-checking? If not, can we just delete the parse rules for now?

Reply:
Per Leo's comment, it seems better to reject in type-checking rather than deleting the parse rules.